AppDashboard/lib/app_dashboard_data.py

```diff
@@ -647,3 +735,152 @@ def can_upload_apps(self):
+      except Exception as err:
+        logging.exception(err)
+        pass
+    if user_info:
+      if type(values) is not dict:
+        #base admin template
+        values = {
+          "nav":["app_management", "appscale_management",
+          "debugging_monitoring"],
+          "panel":["app_console","upload_app","cloud_stats","database_stats",
+          "memcache_stats"]
+          }
+      layout_settings=values
+      LOOKUP_DICT = self.build_dict(user_info=user_info)
+      layout_settings['nav'] = [{key: LOOKUP_DICT.get(key)} for key in
+                                layout_settings.get('nav') if
+                                key in LOOKUP_DICT]
+
+      layout_settings['panel'] = [{key: LOOKUP_DICT.get(key)} for key in
+                                  layout_settings.get('panel') if
+                                  key in LOOKUP_DICT and (LOOKUP_DICT.get(key).get(
+                                  'is_admin_panel') == user_info.is_user_cloud_admin
+                                  or not LOOKUP_DICT.get(key).get('is_admin_panel'))]
+      user_info.dash_layout_settings = layout_settings
+      user_info.put()
+      return user_info.dash_layout_settings
+    return
+
+
+  def rebuild_dash_layout_settings_dict(self, email=None):
+    """ Rebuilds the user's layout settings in case there is an update
+      to the lookup dictionary
+
+    Args:
+      email: A str that indicates the e-mail address of the user logging in
+    """
+    if email is None:
+      return {}
+    try:
+      user_info = self.get_by_id(UserInfo, email)
+      if user_info:
+        try:
+          if user_info.dash_layout_settings:
+            LOOKUP_DICT = self.build_dict(user_info=user_info)
+            values = user_info.dash_layout_settings
+            default_nav = ["app_management","appscale_management",
+                           "debugging_monitoring"]
+
+            nav_list=[]
+            for key_dict in values.get('nav'):
+              for temp_key in key_dict:
+                nav_list.append(temp_key)
+
+            if set(nav_list)!=set(default_nav):
+              for key in default_nav:
+                if nav_list.count(key)==0:
+                  nav_list.append(key)
+
+            default_panel = ["app_console","upload_app","cloud_stats",
+                             "database_stats","memcache_stats"]
+
+            panel_list=[]
+            for key_dict in values.get('panel'):
+              for temp_key in key_dict:
+                panel_list.append(temp_key)
+
+            if set(panel_list)!= set(default_panel):
+              for key in default_panel:
+                if panel_list.count(key)==0:
+                  panel_list.append(key)
+
+            values['nav'] = [{key: LOOKUP_DICT.get(key)} for key in
+                                nav_list if
+                                key in LOOKUP_DICT]
+
+            values['panel'] = [{key: LOOKUP_DICT.get(key)} for key in
```

Comment:
Don't be afraid to use local variables for large accessors like `LOOKUP_DICT.get(key).get('is_admin_panel')`.